lib/auth/result.rb

```diff
@@ -165,7 +156,27 @@ def to_client_hash
 
   private
 
+  def staged_user
+    @staged_user ||= begin
+      if email.present? && email_valid
+        User.where(staged: true).find_by_email(email)
+      else
+        nil
+      end
+    end
+  end
+
   def username_suggester_attributes
     username || name || email
   end
+
+  def resolve_username
+    if staged_user
+      if !username.present? || UserNameSuggester.fix_username(username) == staged_user.username
+        return staged_user.username
+      end
+    end
```

Comment:
Feels a little weird to nest two `if` statements like this - could we combine them? 🤔 

```suggestion
    if staged_user && username.present? && UserNameSuggester.fix_username(username) == staged_user.username
      return staged_user.username
    end
```

Reply:
I knew you'd pay attention to it :).

I consciously wrote it like this, since I believe this version is much more clean and readable.

I'm ok with changing it, but at the same time I would love not only to leave it like this but also to see us doing the same things a bit more often in our codebase.

Normally, when you solve some task you need to read a lot of small pieces of code, sometimes it's dozens of methods. We tend to always combine such conditions in one `if` statement, which makes our codebase somewhat harder to scan.

For example, I'd scan my version of code like this:

![2021-12-03 22 33 24](https://user-images.githubusercontent.com/1274517/144675659-3e2b33cd-188f-4098-b729-9c02af49c571.jpg)

I'd go to the method and I see, aha, ok, we special case staged users, and in other cases we use `UserNameSuggester`. And only _sometimes_ I'm going to need to concentrate on the inner `if` and look what exactly we do for staged users.

It's like in writing. A text that written in short sentences is easier to read and understand. A text with good paragraphs and headers is easier to understand too.

And this is like a header:
```ruby
if staged_user
...
end
```

It makes my intentions super clear - I special case staged users here.

Your version looks more like "If the user is staged, and we have username in payload and that username is the same as the username of staged user then...".

Reply:
I am persuaded - sounds good to me :+1: